### PR TITLE
[DOC] Installation and How to Run spark-api in README

### PR DESCRIPTION
**Pull request by mcarmonaa**
This PR is related to https://github.com/src-d/spark-api/issues/34.

A first approach to a explanation on how to install and run spark-api, pyspark examples should be added too. Feedback is welcome!

### CODE REVIEW

**Comment by bzz** [comment]
@mcarmonaa Just FYI - PRs are a form of team communication, so usually it's a good engineering practice not to leave the description empty.
   
It usually a good idea to put short description and a link to the appropriate issue, so people who might have focus on other tasks but are asked to review it, could recall what is it all about and find more context about the history of this task.

**Comment by bzz** [comment]
On the content of README, what do you think about this structure

// first line with header \w CI badge (like [all other projects](https://raw.githubusercontent.com/bblfsh/client-scala/master/README.md) other projects
# TL;DR
 - wget ..
 - tar xzf ... ; export
 - spark-shell ....

# Pre-requests
  have Spark installed
  //more details

# Examples of API usage
## pySpark 
local mode
```
 pip intstall ..
 pyspark
```

cluster mode
```
 pyspark --jars ./path/to/spark-api-jar-inside-the-spark-api-python-lib --py-files ./libs.zip
```
This would be the main part of the product
 

## Scala API
```scala
spark-shell --jar ...
import tech.sourced.api._

val api = SparkAPI(spark, path='/path/to/siva/files')
// includes 
//  some more useful imports, if need
//  setConf(, path)
//  something.registrUDFs()


api.getRepositories()
//...

```

File: README.md [hunk 1]
```diff
@@ -1 +1,70 @@
...
+
+```bash
+$ export PATH=$PATH:$HOME/spark-2.2.0-bin-hadoop2.7/bin
+```
+
+
+Run spark-api
+---------------------
+
+By the moment, `spark-api`  can be downloaded from [jitpack](https://jitpack.io), so you should be able to run the `spark-shell` with `spark-api` as a required dependency it the following way:
```
Comment (erizocosmico) [inline README.md]: s/By the moment/For the moment/

Comment (erizocosmico) [reply README.md]: s/can be downloaded/can only be installed/

I would also mention that in the future we're planning to upload the library to maven central

Comment (erizocosmico) [reply README.md]: s/it the following way/in the following way/

File: README.md [hunk 2]
```diff
@@ -1 +1,70 @@
...
+```
+
+To start using spark-api from the shell you must import `spark-api` package and Implicits object from it:
+
+```bash
+scala> import tech.sourced.api._
+import tech.sourced.api._
+
+scala> import Implicits._
+import Implicits._
```
Comment (erizocosmico) [inline README.md]: can't this be simplified to `import tech.sourced.api.Implicits._`?

File: README.md [hunk 3]
```diff
@@ -1 +1,70 @@
...
+```
+
+
+Example of use
+-----------------------
+
+To load siva files as the data source, you have to point to the directory that contains them:
+
+```bash
+scala> spark.sqlContext.setConf("tech.sourced.api.repositories.path", "/path/to/siva-files")
```
Comment (erizocosmico) [inline README.md]: @bzz @ajnavarro @mcarmonaa what do you think if we add an implicit method to the spark context that is like `setRepositoriesPath("path")`? Having to put the key by hand every time (which is a bit long) is error prone

It's not for this PR specifically, but don't know if worth it to raise an issue

Comment (mcarmonaa) [reply README.md]: I've been talked with @bzz about write a function that set all the needed things just like python-wrapper does, Including this `setConf` but in a separate PR

File: README.md (hunk 2 above)
Comment (mcarmonaa) [reply README.md]: Indeed, it can!

File: README.md (hunk 3 above)
Comment (erizocosmico) [reply README.md]: I've done a PR for that in https://github.com/src-d/spark-api/pull/45 so we'd better followup the discussion there (or an issue or somewhere)

**Comment by mcarmonaa** [comment]
According with @bzz proposal, the structure of the documented has been changed. Also changes pointed by @erizocosmico has been addressed.

File: README.md (hunk 3 above)
Comment (mcarmonaa) [reply README.md]: Great!

File: README.md [hunk 4]
```diff
@@ -1 +1,150 @@
-# spark-api
\ No newline at end of file
+# spark-api [![Build Status](https://travis-ci.org/src-d/spark-api.svg?branch=master)](https://travis-ci.org/src-d/spark-api)
+
+
+# TL;DR
+
+Install `spark`:
```
Comment (bzz) [inline README.md]: `spark` -> [Apache Spark](https://spark.apache.org/)

File: README.md [hunk 5]
```diff
@@ -1 +1,150 @@
...
+
+```bash
+$ cd /tmp && wget http://ftp.cixug.es/apache/spark/spark-2.2.0/spark-2.2.0-bin-hadoop2.7.tgz
```
Comment (bzz) [inline README.md]: Let's replace particular mirror URL with `https://www.apache.org/dyn/mirrors/mirrors.cgi?action=download&filename=spark/spark-2.2.0/spark-2.2.0-bin-hadoop2.7.tgz`

File: README.md [hunk 6]
```diff
@@ -1 +1,150 @@
...
+$ tar -C ~/ -xvzf spark-2.2.0-bin-hadoop2.7.tgz
+$ export SCALA_HOME=$HOME/spark-2.2.0-bin-hadoop2.7/bin
```
Comment (bzz) [inline README.md]: `SCALA_HOME` -> `SPARK_HOME`

File: README.md [hunk 7]
```diff
@@ -1 +1,150 @@
...
+```
+
+Run `spark-shell` with `spark-api`:
+
+```bash
+$ $SPARKHOME/bin/spark-shell --packages "com.github.src-d:spark-api:master-SNAPSHOT" --repositories "https://jitpack.io"
```
Comment (bzz) [inline README.md]: SPARKHOME -> SPARK_HOME

File: README.md [hunk 8]
```diff
@@ -1 +1,150 @@
-# spark-api
\ No newline at end of file
+# spark-api [![Build Status](https://travis-ci.org/src-d/spark-api.svg?branch=master)](https://travis-ci.org/src-d/spark-api)
+
+
+# TL;DR
```
Comment (bzz) [inline README.md]: How about changing `TL;DR` to a `Quickstat` and keep all this section in a single MD fenced block, as just a series of commands without explanation? 

For the sake of simplicity, we can also assume everything is done inside SparkAPI clone directory, without `/tmp` and `SPARK_HOME` in this section.

```
wget 
tar xvz
cd spark
/bin/spark-shell ...

#or 

/bin/pyspark ...

```

Keeping all the details for the next section (as it already has).
What do you think?

File: README.md [hunk 9]
```diff
@@ -1 +1,150 @@
...
+Run `pyspark` with `spark-api`:
+
+```bash
+$ $SPARK_HOME/bin/pyspark --repositories "https://jitpack.io"  --packages "tech.sourced:spark-api:master-SNAPSHOT"
+```
+
+
+# Pre-requests
+
+## Spark Installation
```
Comment (bzz) [inline README.md]: Spark -> Apache Spark

File: README.md [hunk 10]
```diff
@@ -1 +1,150 @@
...
+# Examples of API usage
+
+## pySpark
+
+### Local mode
+
+Install python-wrappers is necessary to use spark-api from pyspark:
+
+``` bash
+$ pip install <spark-api>
```
Comment (bzz) [inline README.md]: `<spark-api>` -> same URL as below

File: README.md [hunk 11]
```diff
@@ -1 +1,150 @@
...
+Then you should point to the remote repository where spark-api is hosted and provide the maven coordinates:
+```bash
+$ $SPARK_HOME/bin/pyspark --repositories "https://jitpack.io"  --packages "tech.sourced:spark-api:0.1.0-SNAPSHOT"
+```
+
+### Cluster mode
+
+Install spark-api wrappers as in local mode:
+```bash
+$ sudo pip install -e 'git+https://github.com/erizocosmico/spark-api.git@feature/python-wrapper#egg=spark-api&subdirectory=python'
```
Comment (bzz) [inline README.md]: let's remove `sudo` as it depends on the environment configuration

File: README.md [hunk 12]
```diff
@@ -1 +1,150 @@
...
+```
+
+Then you should package and compress with `zip tool`  the python-wrappers to provide pyspark with it. It's required to distribute the code among the nodes of the cluster.
```
Comment (bzz) [inline README.md]: zip tool -> zip

File: README.md [hunk 13]
```diff
@@ -1 +1,150 @@
...
++--------------------+---------------+--------------------+
+|github.com/mawag/...|refs/heads/HEAD|fff7062de8474d10a...|
++--------------------+---------------+--------------------+
+
+
+```
+
+## Scala API
+
+For the moment, `spark-api`  can only be installed from [jitpack](https://jitpack.io) (coming soon availabe on maven central), so you should be able to run the `spark-shell` with `spark-api` as a required dependency in the following way:
```
Comment (bzz) [inline README.md]: `coming soon availabe on maven central` -> `will be available from the Maven Central soon`

**Comment by bzz** [comment]
@mcarmonaa one more thing - sorry if I have missed it, but as discussed, a section on development should be added similar to i.e https://github.com/src-d/borges#development

Main reason is that test, after #41 will depend on having bblfsh server running, so we should have a
`docker run -d --privileged -p 9432:9432 --name bblfsh bblfsh/server bblfsh server --log-level debug` and a link to https://doc.bblf.sh/

The commands to build a Docker image \w Jupyter notebook from our Docker would be nice to have there, as well as convenience like `./sbt dependencyTree`

File: README.md [hunk 14]
```diff
@@ -1 +1,174 @@
-# spark-api
\ No newline at end of file
+# spark-api [![Build Status](https://travis-ci.org/src-d/spark-api.svg?branch=master)](https://travis-ci.org/src-d/spark-api)
+
+High-level Spark API for running scalable data retrieval pipelines that process and manipulate any number of code repositories for source code analysis. Written mostly in Scala, it aims to be robust, friendly and flexible: it is built on top of Apache Spark, accessible both via Scala and Python Spark APIs, and capable of running on large-scale distributed clusters over petabytes of data.
+
+# Quickstart
+
+```bash
+$ wget https://www.apache.org/dyn/mirrors/mirrors.cgi?action=download&filename=spark/spark-2.2.0/spark-2.2.0-bin-hadoop2.7.tgz
```
Comment (bzz) [inline README.md]: We need to put url inside the `""` otherwise if copied, might not work, depending on the shell

File: README.md [hunk 15]
```diff
@@ -1 +1,174 @@
-# spark-api
\ No newline at end of file
+# spark-api [![Build Status](https://travis-ci.org/src-d/spark-api.svg?branch=master)](https://travis-ci.org/src-d/spark-api)
+
+High-level Spark API for running scalable data retrieval pipelines that process and manipulate any number of code repositories for source code analysis. Written mostly in Scala, it aims to be robust, friendly and flexible: it is built on top of Apache Spark, accessible both via Scala and Python Spark APIs, and capable of running on large-scale distributed clusters over petabytes of data.
```
Comment (bzz) [inline README.md]: High-level Spark API is for ...

File: README.md [hunk 16]
```diff
@@ -1 +1,174 @@
...
+# or
+$ bin/pyspark --repositories "https://jitpack.io"  --packages "com.github.src-d:spark-api:master-SNAPSHOT"
+```
+
+
+# Pre-requests
+
+## Apache Spark Installation
+
+Firstly, you need to download [Apache Spark](https://spark.apache.org/) somewhere on your machine:
```
Comment (bzz) [inline README.md]: Firstly -> First

File: README.md [hunk 17]
```diff
@@ -1 +1,174 @@
...
\ No newline at end of file
+# spark-api [![Build Status](https://travis-ci.org/src-d/spark-api.svg?branch=master)](https://travis-ci.org/src-d/spark-api)
+
+High-level Spark API for running scalable data retrieval pipelines that process and manipulate any number of code repositories for source code analysis. Written mostly in Scala, it aims to be robust, friendly and flexible: it is built on top of Apache Spark, accessible both via Scala and Python Spark APIs, and capable of running on large-scale distributed clusters over petabytes of data.
+
+# Quickstart
+
+```bash
+$ wget https://www.apache.org/dyn/mirrors/mirrors.cgi?action=download&filename=spark/spark-2.2.0/spark-2.2.0-bin-hadoop2.7.tgz
+$ tar -xzf spark-2.2.0-bin-hadoop2.7.tgz
```
Comment (bzz) [inline README.md]: Add `; cd spark-2.2.0-bin-hadoop2.7` to the end of this line.

This would make next 2 lines work as expected

File: README.md [hunk 18]
```diff
@@ -1 +1,174 @@
...
+```
+
+# Development
+
+## Build fatjar
+
+Build the fatjar is needed to build the docker image that contains the jupyter server,  or test changes in spark-shell just passing the jar with `--jars` flag:
+
+```bash
+$ sbt assembly
```
Comment (bzz) [inline README.md]: here and below `sbt` -> `./sbt`

File: README.md [hunk 19]
```diff
@@ -1 +1,178 @@
...
+$ wget "https://www.apache.org/dyn/mirrors/mirrors.cgi?action=download&filename=spark/spark-2.2.0/spark-2.2.0-bin-hadoop2.7.tgz"
+$ tar -xzf spark-2.2.0-bin-hadoop2.7.tgz; cd spark-2.2.0-bin-hadoop2.7
+$ bin/spark-shell --packages "com.github.src-d:spark-api:master-SNAPSHOT" --repositories "https://jitpack.io"
+
+# or
+$ bin/pyspark --repositories "https://jitpack.io"  --packages "com.github.src-d:spark-api:master-SNAPSHOT"
+```
+
+
+# Pre-requests
```
Comment (erizocosmico) [inline README.md]: s/requests/requisites/

**Comment by bzz** [comment]
Looks awesome.

What do you think if we merge it now, and then submit further minor improvements on top of that?

File: README.md [hunk 20]
```diff
@@ -1 +1,178 @@
...
+# spark-api [![Build Status](https://travis-ci.org/src-d/spark-api.svg?branch=master)](https://travis-ci.org/src-d/spark-api)
+
+High-level Spark API is for running scalable data retrieval pipelines that process and manipulate any number of code repositories for source code analysis. Written mostly in Scala, it aims to be robust, friendly and flexible: it is built on top of Apache Spark, accessible both via Scala and Python Spark APIs, and capable of running on large-scale distributed clusters over petabytes of data.
+
+# Quickstart
+
+```bash
+$ wget "https://www.apache.org/dyn/mirrors/mirrors.cgi?action=download&filename=spark/spark-2.2.0/spark-2.2.0-bin-hadoop2.7.tgz"
+$ tar -xzf spark-2.2.0-bin-hadoop2.7.tgz; cd spark-2.2.0-bin-hadoop2.7
+$ bin/spark-shell --packages "com.github.src-d:spark-api:master-SNAPSHOT" --repositories "https://jitpack.io"
```
Comment (erizocosmico) [inline README.md]: doesn't it need also enry locally for now?

File: README.md [hunk 21]
```diff
@@ -1 +1,178 @@
...
+```bash
+$ tar -C ~/ -xvzf spark-2.2.0-bin-hadoop2.7.tgz
+```
+Binaries and scripts to run `Spark` are located in spark-2.2.0-bin-hadoop2.7/bin, so maybe you would like to add it to your `PATH`:
+
+```bash
+$ export PATH=$PATH:$HOME/spark-2.2.0-bin-hadoop2.7/bin
+```
+
+or just set SPARK_HOME and run it as following:
```
Comment (erizocosmico) [inline README.md]: s/SPARK_HOME/`SPARK_HOME`/

File: README.md [hunk 22]
```diff
@@ -1 +1,178 @@
...
+
+```bash
+$ export SPARK_HOME=$HOME/spark-2.2.0-bin-hadoop2.7
+$ $SPARK_HOME/bin/spark-shell
+```
+
+# Examples of API usage
+
+## pySpark
```
Comment (erizocosmico) [inline README.md]: s/pySpark/pyspark/

File: README.md [hunk 23]
```diff
@@ -1 +1,178 @@
...
+
+### Local mode
+
+Install python-wrappers is necessary to use spark-api from pyspark:
+
+``` bash
+$ pip install  'git+https://github.com/erizocosmico/spark-api.git@feature/python-wrapper#egg=spark-api&subdirectory=python'
```
Comment (erizocosmico) [inline README.md]: We should change this to be the real URL since this is likely to be merged very soon

File: README.md [hunk 24]
```diff
@@ -1 +1,178 @@
...
+Then you should point to the remote repository where spark-api is hosted and provide the maven coordinates:
+```bash
+$ $SPARK_HOME/bin/pyspark --repositories "https://jitpack.io"  --packages "tech.sourced:spark-api:0.1.0-SNAPSHOT"
+```
+
+### Cluster mode
+
+Install spark-api wrappers as in local mode:
+```bash
+$ pip install -e 'git+https://github.com/erizocosmico/spark-api.git@feature/python-wrapper#egg=spark-api&subdirectory=python'
```
Comment (erizocosmico) [inline README.md]: same

File: README.md [hunk 25]
```diff
@@ -1 +1,178 @@
...
+```
+
+Then you should package and compress with `zip`  the python-wrappers to provide pyspark with it. It's required to distribute the code among the nodes of the cluster.
```
Comment (erizocosmico) [inline README.md]: s/python-wrappers/python wrapper/

File: README.md [hunk 26]
```diff
@@ -1 +1,178 @@
...
+
+```bash
+$ zip <path-to-installed-package> ./spark-api.zip
+$ $SPARK_HOME/bin/pyspark <same-args-as-local-plus> --py-files ./spark-api.zip
+```
+
+### pySpark API usage
```
Comment (erizocosmico) [inline README.md]: s/pySpark/pyspark/

File: README.md [hunk 27]
```diff
@@ -1 +1,178 @@
...
++--------------------+---------------+--------------------+
+|github.com/mawag/...|refs/heads/HEAD|fff7062de8474d10a...|
++--------------------+---------------+--------------------+
+
+
+```
+
+## Scala API
+
+For the moment, `spark-api`  can only be installed from [jitpack](https://jitpack.io) (will be available from the Maven Central soon), so you should be able to run the `spark-shell` with `spark-api` as a required dependency in the following way:
```
Comment (erizocosmico) [inline README.md]: s/from the/from/

File: README.md [hunk 28]
```diff
@@ -1 +1,178 @@
...
+
+```bash
+$ spark-shell --packages com.github.src-d:spark-api:master-SNAPSHOT --repositories https://jitpack.io
+```
+
+To start using spark-api from the shell you must import `spark-api` package and Implicits object from it:
+
+```bash
+scala> import tech.sourced.api.Implicits_
+import tech.sourced.api.Implicits_
```
Comment (erizocosmico) [inline README.md]: s/Implicits_/Implicits._/ (the dot between `Implicits` and `_`)

File: README.md [hunk 29]
```diff
@@ -1 +1,178 @@
...
+spark-api uses [bblfsh](https://github.com/bblfsh) so you need an instance of a bblfsh server running, and you can get one so easy with docker:
+
+```bash
+docker run -d --privileged -p 9432:9432 --name bblfsh bblfsh/server bblfsh server --log-level debug
+```
+
+To run tests:
+```bash
+$ ./sbt tests
+```
```
Comment (erizocosmico) [inline README.md]: Maybe we can add also how to run the python tests.

```
cd python
python -m unittest discover -v
```

File: README.md [hunk 30]
```diff
@@ -1 +1,178 @@
...
+
+# License
+
+Apache License Version 2.0, see [LICENSE](https://github.com/src-d/spark-api/blob/master/LICENSE    )
```
Comment (erizocosmico) [inline README.md]: s/LICENSE(spaces))/LICENSE)/

File: README.md (hunk 23 above)
Comment (bzz) [reply README.md]: Shall we take care of changing URL in the same PR that introduces this feature?
If current one works

File: README.md (hunk 20 above)
Comment (bzz) [reply README.md]: Right.. As we use `-uber` classifier for `./sbt assembly` jar name, we are not getting fatJar from Jitpack and thus, need to specify all dependencies: jgit, bblfhs, siva, etc

How about fixing this in sub-sequent PRs?

Comment (erizocosmico) [reply README.md]: Sounds good

**Comment by mcarmonaa** [comment]
I added a subction in development section to explain how to get java-enry in local

**Comment by bzz** [comment]
👍 Awesome work!

Shall we :shipit:  ?
